xarray/core/duck_array_ops.py

```diff
@@ -169,29 +183,38 @@ def cumulative_trapezoid(y, x, axis):
 
 
 def astype(data, dtype, **kwargs):
-
+    if hasattr(data, "__array_namespace__"):
+        xp = get_array_namespace(data)
+        return xp.astype(data, dtype, **kwargs)
     return data.astype(dtype, **kwargs)
 
 
 def asarray(data, xp=np):
     return data if is_duck_array(data) else xp.asarray(data)
 
 
-def as_shared_dtype(scalars_or_arrays):
+def as_shared_dtype(scalars_or_arrays, xp=np):
     """Cast a arrays to a shared dtype using xarray's type promotion rules."""
 
     if any(isinstance(x, cupy_array_type) for x in scalars_or_arrays):
         import cupy as cp
 
         arrays = [asarray(x, xp=cp) for x in scalars_or_arrays]
+    elif any(hasattr(x, "__array_namespace__") for x in scalars_or_arrays):
+        xp = [x for x in scalars_or_arrays if hasattr(x, "__array_namespace__")][
+            0
+        ].__array_namespace__()
+        arrays = [asarray(x, xp=xp) for x in scalars_or_arrays]
+        out_type = dtypes.result_type(*arrays)
+        return [xp.astype(x, out_type, copy=False) for x in arrays]
     else:
-        arrays = [asarray(x) for x in scalars_or_arrays]
+        arrays = [asarray(x, xp=xp) for x in scalars_or_arrays]
```

Comment:
this looks pretty messy, and I don't really understand why we would need to cast all arrays/scalars to the same array type. Is that required by `cupy`?

If not, would it be possible to change this function to something like:
```python
def as_shared_dtype(scalars_or_arrays):
    arrays = [asarray(x) for x in scalars_or_arrays]
    out_type = dtypes.result_type(*arrays)
    return [astype(x, out_type, copy=False) for x in arrays]
```
and then figure out if/how `np.result_type` (in `dtypes.result_type`) can understand the different arrays?

I'm guessing we'd need to somehow figure out how to keep the `cupy` workaround until that implements the Array API in all versions we care for, but it would make this function much easier to read.

This is probably more a question for @pydata/xarray, though.